### PR TITLE
Fitting results matrix

### PR DESCRIPTION
**Pull request by helenst**
Fitting results matrix

This is for #278 

* One per fitting spec, with models on rows and datasets on columns
* Javascript adapted to row/column rather than model/protocol so it can generically handle fitting or simulation experiments (and columns need not have versions)
* Click cell goes through to fitting results form (rather than starting an experiment directly, as with the original matrix)

I've removed the old hacky UI elements for showing fittings in a matrix, although the relevant database fields are still in place.

### CODE REVIEW

**Review by jonc125 — CHANGES_REQUESTED** [review]
Just been having a quick play and had a couple of ideas for improvement; I'll look through properly tomorrow.

1. It would be good to restrict the columns so they only show datasets linked to protocols that are linked to the fitting specification. Otherwise you get potentially many blank cells, and clicking on them leads to a fitting setup page that's "stuck" as the pre-selections aren't actually allowed!
2. The matrix page should link back to the fitting specification - perhaps in the title where its name is mentioned?

File: weblab/fitting/tests/test_views.py [hunk 1]
```diff
@@ -942,3 +978,530 @@ def test_fittingspec_transfer_success(self, client, logged_in_user, other_user,
...
+        """
+        This is a dumb page that doesn't actually load any data, so we just
+        check that the URLs are working.
+        """
+        response = client.get(url % public_fittingspec.id)
+        assert response.status_code == 200
+
+    def test_matrix_page_for_fitting_spec(self, client, public_fittingspec):
+        response = client.get('/fitting/specs/%d/results' % public_fittingspec.pk)
+        assert response.status_code == 200
```
Comment (jonc125) [inline weblab/fitting/tests/test_views.py:1010]:
```suggestion
```
Would this be a sensible reduction in number of tests?

File: weblab/fitting/tests/test_views.py [hunk 2]
```diff
@@ -942,3 +978,530 @@ def test_fittingspec_transfer_success(self, client, logged_in_user, other_user,
...
+        assert response.status_code == 404
+
+
+@pytest.mark.django_db
+class TestFittingSpecResultsMatrixJsonView:
+    def test_matrix_json_for_fitting_spec(self, client, public_fittingspec):
+        response = client.get('/fitting/specs/%d/results/matrix' % public_fittingspec.pk)
+        assert response.status_code == 200
+        data = json.loads(response.content.decode())
+        assert 'getMatrix' in data
```
Comment (jonc125) [inline weblab/fitting/tests/test_views.py:1024]: Isn't this redundant with `test_matrix_json` below?

File: weblab/fitting/tests/test_views.py [hunk 3]
```diff
@@ -942,3 +978,530 @@ def test_fittingspec_transfer_success(self, client, logged_in_user, other_user,
...
+        unrelated_dataset = recipes.dataset.make(protocol=unrelated_protocol)
+
+        fittingspec = recipes.fittingspec.make(protocol=related_protocol)
+        helpers.add_version(fittingspec, visibility='public')
+
+        response = client.get('/fitting/specs/%d/results/matrix?subset=all' % fittingspec.pk)
+        data = json.loads(response.content.decode())
+
+        columns = data['getMatrix']['columns']
+        assert len(columns) == 1
```
Comment (jonc125) [inline weblab/fitting/tests/test_views.py:1080]: Maybe make this the third assert so it's easier to see why it fails if it does?

File: weblab/fitting/tests/test_views.py [hunk 4]
```diff
@@ -942,3 +978,530 @@ def test_fittingspec_transfer_success(self, client, logged_in_user, other_user,
...
+        response = client.get('/fitting/specs/%d/results/matrix?subset=all' % fit.fittingspec.pk)
+        data = json.loads(response.content.decode())
+        assert 'getMatrix' in data
+        assert len(data['getMatrix']['rows']) == 0
+        assert len(data['getMatrix']['columns']) == 1
+        assert len(data['getMatrix']['experiments']) == 0
+
+    def test_ignores_unrelated_datasets(self, client, helpers):
+        related_protocol = recipes.protocol.make()
+        helpers.add_version(related_protocol, visibility='public')
```
Comment (jonc125) [inline weblab/fitting/tests/test_views.py:1066]: We can probably get away with `add_fake_version` here (and most/all places below?) to make the tests faster.

**Review by jonc125 — APPROVED** [review]
I've made a few minor suggestions, but overall I'm very happy with this - it seems to work well. So will be good to get some user feedback once deployed :)  I wonder if we should remove the 'fixes' keyword in the description before merging this so the original issue remains open to capture feedback from Michael & Gary there?

**Comment by helenst** [comment]
OK - I've fixed the points you raised and removed the 'fixes' keyword @jonc125

**Pull request merged**